draft-ietf-quic-tls.md

```diff
@@ -1007,144 +813,94 @@ encrypted_pn = ChaCha20(pn_key, counter, nonce, packet_number)
+MAY wish to apply additional restrictions on what data it sends prior to the
+completion of the TLS handshake.  A client otherwise treats 0-RTT keys as
+equivalent to 1-RTT keys, except that it ACKs MUST only be sent with
+1-RTT keys.
 
-Using distinct packet types during the handshake for handshake messages, 0-RTT
-data, and 1-RTT data ensures that the server is able to distinguish between the
-different keys used to remove packet protection.  All of these packets can
-arrive concurrently at a server.
+A client that receives an indication that its 0-RTT data has been accepted by a
+server can send 0-RTT data until it receives all of the server's handshake
+messages.  A client SHOULD stop sending 0-RTT data if it receives an indication
+that 0-RTT data has been rejected.
 
-A server might choose to retain 0-RTT packets that arrive before a TLS
-ClientHello.  The server can then use those packets once the ClientHello
-arrives.  However, the potential for denial of service from buffering 0-RTT
-packets is significant.  These packets cannot be authenticated and so might be
-employed by an attacker to exhaust server resources.  Limiting the number of
-packets that are saved might be necessary.
+A server MUST NOT use 0-RTT keys to protect packets.
 
-The server transitions to using 1-RTT keys after sending its first flight of TLS
-handshake messages, ending in the Finished.
-From this point, the server protects all packets with 1-RTT
-keys.  Future packets are therefore protected with 1-RTT keys.  Initially, these
-are marked with a KEY_PHASE of 0.
+If a server rejects 0-RTT, then the TLS stream will not include any TLS records
+protected with 0-RTT keys.
 
+: 0-RTT data can be acknowledged by the server as it receives it, but any
+  packets containing acknowledgments of 0-RTT data cannot have packet protection
+  removed by the client until the TLS handshake is complete.  The 1-RTT keys
+  necessary to remove packet protection cannot be derived until the client
+  receives all server handshake messages.
 
-### Retransmission and Acknowledgment of Unprotected Packets
 
-TLS handshake messages from both client and server are critical to the key
-exchange.  The contents of these messages determine the keys used to protect
-later messages.  If these handshake messages are included in packets that are
-protected with these keys, they will be indecipherable to the recipient.
+## Receiving Out-of-Order Protected Frames {#pre-hs-protected}
 
-Even though newer keys could be available when retransmitting, retransmissions
-of these handshake messages MUST be sent in packets protected with handshake
-keys.  An endpoint MUST generate ACK frames for these messages and send them in
-packets protected with handshake keys.
+Due to reordering and loss, protected packets might be received by an endpoint
+before the final TLS handshake messages are received.  A client will be unable
+to decrypt 1-RTT packets from the server, whereas a server will be able to
```

Comment:
Isn't this the other way around? The client can use 1-RTT keys right after sending the CFIN, but the server has to wait for the receipt of the CFIN before it can use 1-RTT keys for sealing packets.

Reply:
This is a statement about what's cryptographically possible, because the keys depend on SFIN but not CFIN.

Reply:
I'd still clarify that, while the server possesses the 1-RTT keys, it's not permitted to use them for receiving yet.  But it can use them to send immediately.

Reply:
I restructured these paragraphs a bit.